src/typegen/render.ts

```diff
@@ -0,0 +1,257 @@
+import { outdent } from 'outdent'
+import * as Path from 'path'
+import { DmmfDocument, DmmfTypes } from '../dmmf'
+import { getTransformedDmmf } from '../dmmf/transformer'
+import { getCrudMappedFields } from '../mapping'
+import { defaultFieldNamingStrategy } from '../naming-strategies'
+import { PaginationStrategy } from '../pagination'
+import { hardWriteFile, hardWriteFileSync } from '../utils'
+
+type Options = {
+  prismaClientPath: string
+  typegenPath: string
+  paginationStrategy: PaginationStrategy
+  nexusPrismaImportId?: string
+}
+
+export function generateSync(options: Options): void {
+  doGenerate(true, options)
+}
+
+export function generate(options: Options): Promise<void> {
+  return doGenerate(false, options)
+}
+
+export function doGenerate(sync: true, options: Options): void
+export function doGenerate(sync: false, options: Options): Promise<void>
+export function doGenerate(sync: boolean, options: Options): void | Promise<void> {
+  const paginationStrategy = options.paginationStrategy
+  const prismaClientImportId =
+    Path.isAbsolute(options.typegenPath) && Path.isAbsolute(options.prismaClientPath)
+      ? Path.relative(Path.dirname(options.typegenPath), options.prismaClientPath)
+      : options.prismaClientPath
+  const dmmf = getTransformedDmmf(options.prismaClientPath)
+  const tsDeclaration = render({
+    dmmf,
+    paginationStrategy,
+    prismaClientImportId,
+    nexusPrismaImportId: options.nexusPrismaImportId,
+  })
+
+  if (sync) {
+    hardWriteFileSync(options.typegenPath, tsDeclaration)
+  } else {
+    return hardWriteFile(options.typegenPath, tsDeclaration)
+  }
+}
+
+export function render(params: {
+  dmmf: DmmfDocument
+  prismaClientImportId: string
+  nexusPrismaImportId?: string
+  paginationStrategy: PaginationStrategy
+}) {
+  return `\
+import * as Typegen from '${params.nexusPrismaImportId ?? 'nexus-prisma/typegen'}'
```

Comment:
A lot of the static stuff is now statically imported from the library itself instead of being re-generated every time. See comment above about some of the implications